File: typescipt/src/language/types/variableDeclarationType.ts


export class VariableDeclarationType extends Node {
   public VariableType VariableType { get; protected set; }

   protected VariableDeclarationType(SourceReference reference) : base(reference) {
   }

   public static parse(type: string, reference: SourceReference): VariableDeclarationType {
     if (reference == null) throw new Error(nameof(reference));

     if (type == Keywords.ImplicitVariableDeclaration) return new ImplicitVariableDeclaration(reference);
     if (TypeNames.Contains(type)) return new PrimitiveVariableDeclarationType(type, reference);

     return new CustomVariableDeclarationType(type, reference);
   }

   public abstract createVariableType(context: IValidationContext): VariableType;
}

File: src/Lexy.Compiler/Language/VariableTypes/VariableDeclarationType.cs
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.VariableTypes;

public abstract class VariableDeclarationType : Node
{
    public VariableType VariableType { get; protected set; }

    protected VariableDeclarationType(SourceReference reference) : base(reference)
    {
    }


    protected abstract VariableType ValidateVariableType(IValidationContext context);

    protected override void Validate(IValidationContext context)
    {
        VariableType = ValidateVariableType(context);
    }
}